qiskit/quantum_info/operators/operator.py

```diff
@@ -198,6 +198,99 @@ def from_label(cls, label):
                 op = op.compose(label_mats[char], qargs=[qubit])
         return op
 
+    def apply_permutation(self, perm: list, front: bool = False):
+        """Modifies operator's data by composing it with a permutation.
+
+        Args:
+            perm (list): permutation pattern, describing which qubits
+                occupy the positions 0, 1, 2, etc. after applying the permutation.
+            front (bool): When set to ``True`` the permutation is applied before the
+                operator, when set to ``False`` the permutation is applied after the
+                operator.
+        Returns:
+            Operator: The modified operator.
+
+        Raises:
+            QiskitError: if the size of the permutation pattern does not match the
+                dimensions of the operator.
+        """
+
+        if not front:
+            # The operator is applied first, the permutation is applied after;
+            # however, in terms of matrices, we compute [P][O].
+
+            if len(perm) != len(self._op_shape.dims_l()):
+                raise QiskitError(
+                    "The size of the permutation pattern does not match dimensions of the operator."
+                )
+
+            inv_perm = np.argsort(perm)
+
+            # shape: inv-permuted on left, original on right
+            shape_l = self._op_shape.dims_l()
+            shape_l = shape_l[::-1]
+            shape_l = tuple(shape_l[x] for x in inv_perm)
+            shape_l = shape_l[::-1]
+            shape_r = self._op_shape.dims_r()
+            split_shape = shape_l + shape_r
+
+            # axes: permuted on left, id on right
+            axes_l = tuple((np.argsort(inv_perm[::-1]))[::-1])
+            axes_r = tuple(
+                self._op_shape._num_qargs_l + x for x in range(self._op_shape._num_qargs_r)
+            )
+            split_axes = axes_l + axes_r
+
+            new_mat = (
+                self._data.reshape(split_shape).transpose(split_axes).reshape(self._op_shape.shape)
+            )
+
+            # updating shape: permuted on left, original on right
+            new_shape_l = self._op_shape.dims_l()
+            new_shape_l = new_shape_l[::-1]
+            new_shape_l = tuple(new_shape_l[x] for x in perm)
+            new_shape_l = new_shape_l[::-1]
+            new_shape_r = self._op_shape.dims_r()
+
+            new_op = Operator(new_mat, input_dims=new_shape_r, output_dims=new_shape_l)
```

Comment:
It's hard to tell which parts of this code are logically shared between the two branches and which parts are different, so it's a bit tricky to read.  I think you could re-organise this to
```python
inv_perm = np.argsort(perm)
if front:
    if len(inv_perm) != len(...):
        ...
    shape_l = ...
    shape_r = ...
    axes_l = ...
    axes_r = ...
    new_shape_l = ...
    new_shape_r = ...
else:
    if len(inv_perm) != len(...):
        ...
    shape_l = ...
    ...
split_shape = shape_l + shape_r
axes_order = axes_l + axes_r
new_matrix = self._data.reshape(split_shape).transpile(axes_order).reshape(self._op_shape.shape)
return Operator(new_matrix, input_dims=new_shape_r, output_dims=new_shape_l)
```

To me at least, this makes the logic clearer (assuming I'm correct) - the difference between `front` and `not front` is more obviously localised just to the construction of the shape / permutation axes, and the actual re-arrangement and construction logic is the same from then on.

Reply:
Done.